packages/nx/src/adapter/ngcli-adapter.ts

```diff
@@ -51,10 +52,115 @@ import {
   CustomHasher,
   Executor,
   ExecutorConfig,
+  ExecutorContext,
   ExecutorsJson,
   TaskGraphExecutor,
 } from '../config/misc-interfaces';
 import { readPluginPackageJson } from '../utils/nx-plugin';
+import { major } from 'semver';
+
+export async function createBuilderContext(
+  builderInfo: {
+    builderName: string;
+    description: string;
+    optionSchema: any;
+  },
+  context: ExecutorContext
+) {
+  require('nx/src/adapter/compat');
+  const fsHost = new NxScopedHost(context.root);
+  const { workspace } = await workspaces.readWorkspace(
+    'angular.json',
+    workspaces.createWorkspaceHost(fsHost)
+  );
+  const architectHost = await getWrappedWorkspaceNodeModulesArchitectHost(
+    workspace,
+    context.root
+  );
+
+  const registry = new schema.CoreSchemaRegistry();
+  registry.addPostTransform(schema.transforms.addUndefinedDefaults);
+  registry.addSmartDefaultProvider('unparsed', () => {
+    // This happens when context.scheduleTarget is used to run a target using nx:run-commands
+    return [];
+  });
+  const { Architect } = require('@angular-devkit/architect');
+
+  const architect: Architect = new Architect(architectHost, registry);
+
+  const rxjsMajor = getInstalledPackageVersionInfo('rxjs').major;
+
+  const validateOptions =
+    rxjsMajor > 6
+      ? async (options: json.JsonObject, builderName: string) => {
+          const { firstValueFrom } = await import('rxjs');
+          return firstValueFrom(
+            architect['_scheduler'].schedule('..validateOptions', [
+              builderName,
+              options,
+            ]).output
+          );
+        }
+      : (options: json.JsonObject, builderName: string) => {
+          return architect['_scheduler']
+            .schedule('..validateOptions', [builderName, options])
+            .output.toPromise();
+        };
+
+  const getProjectMetadata =
+    rxjsMajor > 6
+      ? async (target: Target | string) => {
+          const { firstValueFrom } = await import('rxjs');
+          return firstValueFrom(
+            architect['_scheduler'].schedule('..getProjectMetadata', target)
+              .output
+          );
+        }
+      : (target: Target | string) => {
+          return architect['_scheduler']
+            .schedule('..getProjectMetadata', target)
+            .output.toPromise();
+        };
```

Comment:
We could simplify this and wouldn't need the `getInstalledPackageVersionInfo` function at the bottom of the file or the import from `semver`:
```suggestion
  const { firstValueFrom } = require('rxjs');
  const toPromise = (obs: Observable<any>) =>
    firstValueFrom ? firstValueFrom(obs) : obs.toPromise();

  const validateOptions = (options: json.JsonObject, builderName: string) =>
    toPromise(
      architect['_scheduler'].schedule('..validateOptions', [
        builderName,
        options,
      ]).output
    );

  const getProjectMetadata = (target: Target | string) =>
    toPromise(
      architect['_scheduler'].schedule('..getProjectMetadata', target).output
    );
```

Reply:
I like it